test/drbgtest.c

```diff
@@ -864,13 +865,79 @@ static int test_multi_thread(void)
 }
 #endif
 
+/*
+ * This function is only returns the entropy already added with RAND_add(),
+ * and does not get entropy from the OS.
+ */
+static size_t rand_add_entropy(RAND_DRBG *drbg,
+                               unsigned char **pout,
+                               int entropy, size_t min_len, size_t max_len,
+                               int prediction_resistance)
+{
+    size_t ret = 0;
+
+    if (drbg->pool == NULL)
+        return 0;
+
+    if (drbg->pool->entropy < (size_t)entropy || drbg->pool->len < min_len
+        || drbg->pool->len > max_len)
+        return 0;
+
+    *pout = malloc(drbg->pool->len);
```

Comment:
Why do you use plain `malloc()` and not `OPENSSL_malloc()`?

In fact, it is not necessary at all to duplicate the `drbg->pool->buffer`, because the `get_entropy()` and `cleanup_entropy()` callbacks are in full control of the lifetime of the `drbg->pool`. I created a patch demonstrating this which you can squash into your commit.